src/include/detail/linalg/tdb_io.h

```diff
@@ -43,6 +43,77 @@
 #include "utils/print_types.h"
 #include "utils/timer.h"
 
+namespace {
+
+template <class T>
+std::vector<T> read_vector_helper(
+    const tiledb::Context& ctx,
+    const std::string& uri,
+    size_t start_pos,
+    size_t end_pos,
+    size_t timestamp,
+    bool read_full_vector) {
+  scoped_timer _{tdb_func__ + " " + std::string{uri}};
+
+  tiledb::TemporalPolicy temporal_policy =
+      (timestamp == 0) ? tiledb::TemporalPolicy() :
+                         tiledb::TemporalPolicy(tiledb::TimeTravel, timestamp);
+
+  auto array_ = tiledb_helpers::open_array(
+      tdb_func__, ctx, uri, TILEDB_READ, temporal_policy);
+  auto schema_ = array_->schema();
+
+  using domain_type = int32_t;
+  const size_t idx = 0;
+
+  auto domain_{schema_.domain()};
+
+  auto dim_num_{domain_.ndim()};
+  auto array_rows_{domain_.dimension(0)};
+
+  if (read_full_vector) {
+    if (start_pos == 0) {
```

Comment:
In thinking about this, just like `stop_pos == 0` should have no special meaning as a flag, `stop_pos == 0` should have no special meaning (0 should not be a sentinel).  That is, we might want to read a "half-full" vector -- starting at a given place and reading to the end or starting at the beginning and reading to a given place -- and the "given place" in either case could be zero, which would be the wrong thing since right now it is either both or neither.  We can do that currently (though there is a bug -- if you want to actually start at zero but the array doesn't start at zero). So we shouldn't be using a sentinel value for either `start` or `stop`

Unfortunately, we can't determine which half-full case we want with an overload that has one of `start` or `stop` omitted.

There are four cases:
- array begin to array end -- `read_full_vector == true` in the helper -- at which point `start` and `stop` are sentinel values and "end" is indicating by the overloaded function passing the sentinel values to the helper function.
- array begin to `stop` -- not supported in current PR (previously was)
- `start` to array end -- not supported in current PR (previously was, though a bug for `start == 0`)
- `start` to `stop` -- in the current PR, 0 is a valid value for either `start` or `stop`
where zero is a valid value for `start` or `stop` and array begin or array end (or both) may be zero or may not be zero.  Something like Python's range notation would be really helpful....

While we are waiting for C++ to introduce that notation, we need a way to specify "array begin" or "array end" for either `start` or `stop` -- without using a sentinel.  Or at least not using a sentinel that is a value likely to actually be used.  
* One approach would be to use a sentinel that is unlikely to ever be used, like -MAXINT and MAXINT.  Actually, in that case, collisions would be benign -- meaning "read from the beginning" and "read from this value" are the same.
* Or, we could use `std::optional` -- an empty value would mean from the beginning/end for `start`/`stop`.  The notation from the caller side is kind of indicative of what is desired.
```c++
template <class T> std::vector<T> read_vector(const tiledb::Context& ctx, const std::string& uri, 
std::optional<int32_t> start_pos, std::optional<int32_t> end_pos, std::optional<size_t> timestamp);

auto foo = read_vector(ctx, "a.tdb", 0, {}, {});
auto bar = read_vector(ctx, "b.tdb", {}, {}, {}); // Could still have overload: read_vector(ctx, "b.tdb", {});
```
(Note that I am also suggesting we use optional for timestamp, though maybe default value of zero really is what is intended.)

Note also, and very important, `start` and `stop` should be `int32_t` since that is what the vectors and matrix types assume (we should use an appropriate type alias so that they are all consistent).

The constructors for `tdbMatrix` (et al) have the same issues (but twice as many variables for factorially more combinations).

Reply:
Agree there is certainly more we can improve here, but I would propose to go with the current approach because it is working and is simple. If we find scenarios where it doesn't support our needs, then we can update it then.